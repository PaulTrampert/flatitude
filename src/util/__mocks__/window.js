export default {
  setTimeout: jest.fn(),
  clearTimeout: jest.fn()
};