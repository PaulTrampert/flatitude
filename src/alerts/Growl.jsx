import React from 'react';
import PropTypes from 'prop-types';
import ColorTypes from '../colors/ColorTypes.js';
import window from '../util/window.js';
import classnames from 'classnames';

class Growl extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      expiring: false
    };
  }

  componentDidMount = () => {
    let {
      durationMs,
    } = this.props;
    this.timer = window.setTimeout(this.handleExpire, durationMs);
    this.ref && this.ref.addEventListener('animationend', this.onFadeoutEnd);
  }

  componentWillUnmount = () => {
    this.ref && this.ref.removeEventListener('animationend', this.onFadeoutEnd);
  }

  onFadeoutEnd = (event) => {
    if (event.animationName === 'growl-fadeout') {
      this.props.onExpire(this.props.id);
    }
  }

  handleExpire = () => {
    delete this.timer;
    this.setState({expiring: true});
  }

  handleRef = (ref) => {
    this.ref = ref;
  }

  handleDismiss = () => {
    let {
      id,
      onDismiss
    } = this.props;
    this.timer && window.clearTimeout(this.timer);
    delete this.timer;
    onDismiss(id);
  }

  render() {
    let {
      children,
      type,
    } = this.props;
    let {
      expiring
    } = this.state;
    return (
      <div ref={this.handleRef} className={classnames("growl", type, expiring && 'expiring')}>
        <div className="icon">
        </div>
        <div className="message">
          {children}
        </div>
        <div className="dismiss">
          <button onClick={this.handleDismiss}>
          </button>
        </div>
      </div>
    );
  }
}

Growl.propTypes = {
  id: PropTypes.any.isRequired,
  children: PropTypes.node,
  type: ColorTypes.isRequired,
  onDismiss: PropTypes.func.isRequired,
  onExpire: PropTypes.func.isRequired,
  durationMs: PropTypes.number
};

Growl.defaultProps = {
  durationMs: 10000
};

export default Growl;