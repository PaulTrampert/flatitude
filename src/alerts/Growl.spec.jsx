import React from 'react';
import {shallow} from 'enzyme';
import Growl from './Growl.jsx';
import window from '../util/window.js';

jest.mock('../util/window.js');

describe('Growl', () => {
  let subject;
  let instance;
  let onDismiss;
  let onExpire;

  beforeEach(() => {
    onDismiss = jest.fn();
    onExpire = jest.fn();
    subject = shallow(<Growl
      id={2}
      type='info'
      onDismiss={onDismiss}
      onExpire={onExpire}
    >
      hi there!
    </Growl>, {disableLifecycleMethods: true});

    instance = subject.instance();
    instance.ref = {
      addEventListener: jest.fn(),
      removeEventListener: jest.fn()
    };
  });

  describe('componentDidMount', () => {
    it('sets the timeout', () => {
      instance.componentDidMount();

      expect(window.setTimeout).toHaveBeenCalledWith(instance.handleExpire, 10000);
    });

    it('adds an event listener for animationend to the ref', () => {
      instance.componentDidMount();

      expect(instance.ref.addEventListener).toHaveBeenCalledWith('animationend', instance.onFadeoutEnd);
    });
  });

  describe('componentWillUnmount', () => {
    it('removes the event listener from the ref', () => {
      instance.componentWillUnmount();

      expect(instance.ref.removeEventListener).toHaveBeenCalledWith('animationend', instance.onFadeoutEnd);
    });
  });

  describe('onFadeoutEnd', () => {
    let event;

    beforeEach(() => {
      event = {
        animationName: 'growl-fadeout'
      };
    });

    describe('when the animationName is growl-fadeout', () => {
      it('calls onExpire', () => {
        instance.onFadeoutEnd(event);

        expect(onExpire).toHaveBeenCalledWith(2);
      });
    });

    describe('when the animationName is not growl-fadeout', () => {
      beforeEach(() => {
        event.animationName = 'doop';
      });

      it('does not call onExpire', () => {
        instance.onFadeoutEnd(event);

        expect(onExpire).toHaveBeenCalledTimes(0);
      });
    });
  });

  describe("handleExpire", () => {
    beforeEach(() => {
      instance.timer = {};
    });

    it('deletes timer', () => {
      instance.handleExpire();

      expect(instance.timer).not.toBeDefined();
    });

    it('sets expiring to true', () => {
      instance.handleExpire();

      expect(subject.state().expiring).toBe(true);
    });
  });

  describe('handleRef', () => {
    it("sets the ref", () => {
      let ref = {};

      instance.handleRef(ref);
      expect(instance.ref).toBe(ref);
    });
  });

  describe('handleDismiss', () => {
    let timer;
    beforeEach(() => {
      timer = {};
      instance.timer = timer;
    });

    it('clears any timer', () => {
      instance.handleDismiss();

      expect(window.clearTimeout).toHaveBeenCalledWith(timer);
    });

    it('deletes the timer', () => {
      instance.handleDismiss();

      expect(instance.timer).not.toBeDefined();
    });

    it('calls onDismiss', () => {
      instance.handleDismiss();

      expect(onDismiss).toHaveBeenCalledWith(2);
    });
  });

  describe('render', () => {
    it('binds the dismiss button', () => {
      expect(subject.find('div.dismiss button').props().onClick).toBe(subject.instance().handleDismiss);
    });

    it('renders an info growl', () => {
      expect(subject).toMatchSnapshot();
    });

    it('renders a warning growl', () => {
      subject.setProps({type: 'warning'});
      expect(subject).toMatchSnapshot();
    });

    it('renders a success growl', () => {
      subject.setProps({type: 'success'});
      expect(subject).toMatchSnapshot();
    });

    it('renders a danger growl', () => {
      subject.setProps({type: 'danger'});
      expect(subject).toMatchSnapshot();
    });

    it('renders a primary growl', () => {
      subject.setProps({type: 'primary'});
      expect(subject).toMatchSnapshot();
    });

    it('renders an expiring growl', () => {
      subject.setState({expiring: true});
      expect(subject).toMatchSnapshot();
    });
  });
});