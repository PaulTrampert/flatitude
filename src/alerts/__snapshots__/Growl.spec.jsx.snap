// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Growl render renders a danger growl 1`] = `
<div
  className="growl danger"
>
  <div
    className="icon"
  />
  <div
    className="message"
  >
    hi there!
  </div>
  <div
    className="dismiss"
  >
    <button
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Growl render renders a primary growl 1`] = `
<div
  className="growl primary"
>
  <div
    className="icon"
  />
  <div
    className="message"
  >
    hi there!
  </div>
  <div
    className="dismiss"
  >
    <button
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Growl render renders a success growl 1`] = `
<div
  className="growl success"
>
  <div
    className="icon"
  />
  <div
    className="message"
  >
    hi there!
  </div>
  <div
    className="dismiss"
  >
    <button
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Growl render renders a warning growl 1`] = `
<div
  className="growl warning"
>
  <div
    className="icon"
  />
  <div
    className="message"
  >
    hi there!
  </div>
  <div
    className="dismiss"
  >
    <button
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Growl render renders an expiring growl 1`] = `
<div
  className="growl info expiring"
>
  <div
    className="icon"
  />
  <div
    className="message"
  >
    hi there!
  </div>
  <div
    className="dismiss"
  >
    <button
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Growl render renders an info growl 1`] = `
<div
  className="growl info"
>
  <div
    className="icon"
  />
  <div
    className="message"
  >
    hi there!
  </div>
  <div
    className="dismiss"
  >
    <button
      onClick={[Function]}
    />
  </div>
</div>
`;
